### PR TITLE
Histogram2d implementation

### PR DESCRIPTION
**Pull request by alexminnaar**
Implementing numpy's [histogram2d](https://numpy.org/doc/stable/reference/generated/numpy.histogram2d.html).  It should be a straight-forward application of histogramdd for the 2d case.

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -921,6 +921,19 @@ def histogram(a, bins=10, range=None, weights=None, density=None):
     counts = counts / bin_widths / counts.sum()
   return counts, bin_edges
 
+@_wraps(np.histogram2d)
+def histogram2d(x, y, bins=10, range=None, weights=None, density=None):
+
+  N = len(bins)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:927]: It looks like this would result in a `TypeError` for the default value.

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -2325,6 +2325,38 @@ def testHistogram(self, shape, dtype, bins, density, weights):
...
+    for bins in [[2, 2], [[0, 1, 3, 5], [0, 2, 3, 4, 6]]]
+    for weights in [False, True]
+    for density in [False, True]
+  ))
+  def testHistogram2d(self, shape, dtype, bins, weights, density):
+    rng = jtu.rand_default(self.rng())
+    _weights = lambda w: abs(w) if weights else None
+    np_fun = lambda a, b, w: np.histogram2d(a, b, bins=bins, weights=_weights(w), density=density)
+    jnp_fun = lambda a, b, w: jnp.histogram2d(a, b, bins=bins, weights=_weights(w), density=density)
+    args_maker = lambda: [rng(shape, dtype), rng(shape, dtype), rng((shape[0],), dtype)]
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:2350]: In this context, I believe all three results should be `rng(shape, dtype)`

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -2325,6 +2325,38 @@ def testHistogram(self, shape, dtype, bins, density, weights):
...
+      "shape": shape,
+      "dtype": dtype,
+      "bins": bins,
+      "weights": weights,
+      "density": density
+    }
+    for shape in [(5,), (12,)]
+    for dtype in int_dtypes
+    # We only test explicit integer-valued bin edges because in other cases
+    # rounding errors lead to flaky tests.
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:2340]: This comment doesn't match the code below (`bins=[2, 2]` is specifying a number of bins for each axis, rather than explicit integer-valued bin edges)

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great! Just a couple minor comments.

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (alexminnaar) [reply jax/numpy/lax_numpy.py:927]: Yes you're right

File: tests/lax_numpy_test.py (hunk 2 above)
Comment (alexminnaar) [reply tests/lax_numpy_test.py:2350]: yep

File: tests/lax_numpy_test.py (hunk 3 above)
Comment (alexminnaar) [reply tests/lax_numpy_test.py:2340]: Ah that's right.  I think I also made that mistake [here](https://github.com/google/jax/blob/cf1fefcdc65872786e7ed8aafdade35dd4fe8e6c/tests/lax_numpy_test.py#L2339), should I add that change to this PR?

Comment (jakevdp) [reply tests/lax_numpy_test.py:2340]: Sure, that sounds good.

File: tests/lax_numpy_test.py [hunk 4]
```diff
@@ -2325,6 +2325,36 @@ def testHistogram(self, shape, dtype, bins, density, weights):
...
+      jtu.format_shape_dtype_string(shape, dtype), bins, weights, density),
+      "shape": shape,
+      "dtype": dtype,
+      "bins": bins,
+      "weights": weights,
+      "density": density
+    }
+    for shape in [(5,), (12,)]
+    for dtype in int_dtypes
+    for bins in [[2, 2], [[0, 1, 3, 5], [0, 2, 3, 4, 6]]]
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:2339]: I'd also test `bins=2` to cover the case of passing an integer.

Comment (jakevdp) [inline tests/lax_numpy_test.py:2339]: I'd also test `bins=2` to cover the integer case.

**Comment by jakevdp** [comment]
Looks great, thanks! I'm going to trigger some GPU & TPU tests for this code, and if everything passes we'll get it merged. Thanks for all the work on this, especially through the multiple rounds of review!

File: tests/lax_numpy_test.py (hunk 4 above)
Comment (alexminnaar) [reply tests/lax_numpy_test.py]: Good catch, histogrammdd wasn't handling this case correctly.  The way it should work is if bins is an integer then `bins=D*[bins]`.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
One thing: can you rebase on master, squash all the commits into one and force-push to this branch? Thanks.

Let me know if you need help with that.

**Comment by alexminnaar** [comment]
I'm not sure what that failure is referring to
` jax/experimental/jax2tf/jax2tf.py:221: error: Cannot call function of unknown type`
Is that due to my changes?

**Comment by jakevdp** [comment]
The error looks unrelated to this change.

Try rebasing on master, and then force-push.
